### DISCUSSION TITLE
Calculating IISs

### DISCUSSION THREAD

**Discussion by ghauser8** [ERGO-Code/HiGHS#1996] [Ideas]
Hi Team,

Are there any plans to provide an Irreducible Infeasible Subsystem (IIS) calculator as part of the HiGHS solver?

I've heard one of the big reasons people stick with commercial solvers is for IIS logging and compute capabilities. I've done a cursory search and haven't found any open-source calculators for IISs. 

If HiGHS isn't planning to do this, do you know of/have a recommendation for an open-source IIS solver?

Thanks in advance!

**Comment by jajhall** [comment]
It's already WIP in HiGHS, because I know it's values by practitioners. I started in the summer but ran out of time and also realised that I should do work on the primal simplex solver to support IIS. It's a lot of work to do well, and I know of no open-source solver that offers it. Unfortunately I can't see myself having time to return to it until next summer

**Comment by rschwarz** [comment]
I've heard that SCIP has some support for IIS and have just found the [MinIISC](https://github.com/scipopt/scip/tree/d23dfad30098b323c38ebaf052bc8e4e40191759/applications/MinIISC) application within in, but I'm not sure if it's really performing the same function.

**Reply by jajhall** [reply]
When I searched for open-source IIS capability, I found this, too. Reading [more](https://www.scipopt.org/scip/doc/html/MINIISC_MAIN.php) just now, I see that it does relate to IIS for LP. However, it focuses on finding the IIS of minimum cover, and requires a MIP to be solved. Since it's not clear how to use it to find a single IIS - that only needs hot-started LP solves - it didn't look the be a practical code for this problem. Tellingly, perhaps, it's not listed as a feature of SCIP - when it would be a very valuable tool to offer. Hence I deduce that it's not an "available" technique for what users want to know.

**Reply by rschwarz** [reply]
Sorry for the noise, then. My quick glance suggested that a minimum cover is solved in order to find the IIS...

**Reply by jajhall** [reply]
No problem, your query caused me to go back and find out more about what Pfetsch did for SCIP. An IIS if minimal cover might be satisfying to find from a research point of view, but practitioners would rather have a small IIS found heuristically

**Comment by ghauser8** [comment]
Thanks for the discussion, this is all good info. Glad to see you're way ahead of me!

**Comment by Dysg** [comment]
The inclusion of an Irreducible Infeasible Subsystem (IIS) [calculadora de porcentagem](https://porcentagemcalculadora.com/) in the HiGHS solver would indeed be beneficial, as IIS functionality is a key feature in many commercial solvers. Currently, there’s no official word on HiGHS incorporating this, but it’s worth keeping an eye on updates. For open-source IIS solvers, alternatives like SCIP or COIN-OR's Cbc may offer relevant functionality, though with varying capabilities...

Hope this helps! For more help you should ask a [developer](https://thehillclimbing.com/) to solve this issue.

**Reply by jajhall** [reply]
> For open-source IIS solvers, alternatives like SCIP or COIN-OR's Cbc may offer relevant functionality, though with varying capabilities.

There's no sign of IIS calculation in Cbc, and there was an exchange on the SCIP mailing list this morning about it being implemented for SCIP, but it "will sadly not be as performant as the algorithms implemented in some of the closed-source solvers. A substantial amount of work needs to be done for that to happen."

**Comment by jaredxfeng** [comment]
Any updates on this, beloved HiGHS devs and community?

**Reply by jajhall** [reply]
There is "IIS Lite" that will detect some basic causes of infeasibility - in the next release.

Otherwise, there's a plan to work on IIS infrastructure and enhancements to the primal simplex solver that will improve efficiency. We're hoping to start at the end of this year

**Reply by jaredxfeng** [reply]
Sounds good. Looking forward to it!
